flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.flink.runtime.state.StreamStateHandle;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+public class RocksDbStateDataTransfer {
+
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		downloadDataForAllStateHandles(sstFiles, dest, restoringThreadNum, closeableRegistry);
+		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void downloadDataForAllStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry
+		) throws Exception {
+
+		List<Runnable> runnables = createDownloadRunnables(stateHandleMap, restoreInstancePath, closeableRegistry);
+
+		if (enableMultiThreadDownload(restoringThreadNum)) {
+			runTransferWithMultipleThreads(runnables, restoringThreadNum, closeableRegistry);
+		} else {
+			runTransferWithSingleThread(runnables);
+		}
+	}
+
+	private static List<Runnable> createDownloadRunnables(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		CloseableRegistry closeableRegistry
+		) {
+		List<Runnable> runnables = new ArrayList<>(stateHandleMap.size());
+		for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
+			StateHandleID stateHandleID = entry.getKey();
+			StreamStateHandle remoteFileHandle = entry.getValue();
+
+			Path path = new Path(restoreInstancePath, stateHandleID.toString());
+			runnables.add(() -> downloadDataForStateHandle(path, remoteFileHandle, closeableRegistry));
+		}
+		return runnables;
+	}
+
+	private static void runTransferWithMultipleThreads(
+		List<Runnable> runnables,
+		int threadNum,
+		CloseableRegistry closeableRegistry) throws ExecutionException, InterruptedException, IOException {
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(threadNum);
+		Closeable shutDownAllTasksAndExecutorService = () -> {
+			if (executorService != null) {
+				executorService.shutdownNow();
+			}
+		};
+		closeableRegistry.registerCloseable(shutDownAllTasksAndExecutorService);
```

Comment:
Why do we register the `ExecutorService` at the `CloseableRegistry`?

Reply:
I think this should not be needed.

Reply:
True, if the main thread, restoring from checkpoint, gets interrupted correctly, it will interrupt `FutureUtils.waitForAll` and executorService can be always `shutdownNow` in `finally` block.